docs/code_style.md

```diff
@@ -0,0 +1,235 @@
+
+### P10 constants definition keywords
+
+For class-level constants in .hpp files or constants in .cpp files:
+
+Use `static constexpr` if possible (inline is implicit):
+
+	static constexpr uint64_t kMinDifficulty{0x20000};
+
+Otherwise use `static const`.
+
+For global constants in .hpp files:
+
+Use `static constexpr` if possible (inline is implicit),  
+Otherwise use `inline static const`.
+
+See also: [Constants: Safe Idioms](https://abseil.io/tips/140)
+
+### P11 CLion code inspections
+
+CLion has some extra code inspections in addition to clang-tidy. These inspections are not enforced. It is up to each developer to decide if they are useful or not (and enable/disable them locally).
+
+### P12 string formatting
+
+Use `<<` (and `std::stringstream`) or `+` syntax as you feel for logs and error messages. `std::format` is not supported, but [planned eventually](https://github.com/erigontech/silkworm/issues/2384).
+
+### P13 include paths and quotes
+
+Use double quotes and paths relative to the current file within a CMake library, allow referring ancestor directories within the library:
+
+	#include "types.hpp"
+	#include "../../../api/endpoint/range.hpp"
+
+Otherwise use `<` and paths relative to the silkworm root source directory:
+
+	#include <silkworm/core/common/base.hpp>  // outside core
+
+### P14 io_context vs any_io_executor
+
+Most of the async code needs to dispatch operations (or create sub objects that need to dispatch), and doesn't need to run or stop the io_context. In this case the executor interface is enough. `any_io_executor` is a lightweight copyable and movable type-erased wrapper that can be used to dispatch async work without binding to a concrete executor type (such as `io_context::executor`). Prefer using `any_io_executor` if possible instead of `io_context`.
+
+`any_io_executor` can be passed by value, but clang-tidy usually wants to pass it by a const reference:
+
+	const boost::asio::any_io_executor& executor
+
+### P15 getters in .hpp or .cpp
+
+Put simple single line getters in .hpp:
+
+	int id() const { return id_; }
+
+### P16 i++ vs ++i
+
+Prefer ++i by default. Only use i++ where a previous result value is needed.
+
+Good:
+
+	for (size_t i = 0; i < items.size(); ++i)
+
+Bad:
+
+	for (size_t i = 0; i < items.size(); i++)
+
+See [this guideline](https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement)
+
+### P17 auto for constants
+
+Use an explicit type specification instead of `auto` in constants:
+
+Good:
+
+static constexpr size_t kThreadNameFixedSize = 11;
+
+Bad:
+
+static constexpr auto kThreadNameFixedSize = 11;
+
+### P18a optional for an empty std::function
+
+Sometimes we use `std::function` as a factory or a callback object. If it is optional prefer using `optional<function<T(U)>>` and `nullopt` where null is expected instead of using an empty `std::function` (with a `nullptr` inside).
```

Comment:
Inconsistent with the recommendation for `unique_ptr` below.

Reply:
To my mind it should be the opposite. `std::function` already supports the nil state by design, so I don't see a need to complicate things and use `optional<function<T(U)>>`.

Reply:
Yes, it is inconsistent. It was a team decision on the meeting, after we've considered the alternative you suggest. The rationale for this difference is practical: for `std::function` the majority of usages (>95%) implies nonnull, and we want to be explicit about a few null cases. For `unique_ptr` it was not a clear cut (a lot of nullable cases), and we didn't want to introduce any extra noise, even though a distinction might be useful (as in gsl::not_null).